test/extended/openstack/cpms.go

```diff
@@ -0,0 +1,131 @@
+	var controlPlaneMachineSets []objx.Map
+
+	g.BeforeEach(func() {
+		ctx = context.Background()
+
+		g.By("preparing openshift dynamic client")
+		cfg, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dc, err = dynamic.NewForConfig(cfg)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		clientSet, err = e2e.LoadClientset()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		skipUnlessMachineAPIOperator(ctx, dc, clientSet.CoreV1().Namespaces())
+		controlPlaneMachineSets, err = getControlPlaneMachineSets(ctx, dc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+
+	g.It("has role master", func() {
+		labels := controlPlaneMachineSets[0].Get("spec.template.machines_v1beta1_machine_openshift_io.metadata.labels").ObjxMap()
+		o.Expect(labels[machineLabelRole]).To(o.Equal("master"), "unexpected or absent role label")
+	})
+
+	g.It("replica number corresponds to the number of Machines", func() {
+		replicaNumber, _ := strconv.Atoi(controlPlaneMachineSets[0].Get("spec.replicas").String())
+		o.Expect(getMasterMachines(ctx, dc)).
+			To(o.HaveLen(replicaNumber), "unexpected number of replicas for CPMS")
+	})
+
+	g.It("ProviderSpec template is correctly applied to Machines", func() {
+		masterMachines, err := getMasterMachines(ctx, dc)
+		o.Expect(err).NotTo(o.HaveOccurred(), "error fetching Masters machines")
+
+		providerSpec := controlPlaneMachineSets[0].Get("spec.template.machines_v1beta1_machine_openshift_io.spec.providerSpec").ObjxMap()
+		cpmsFlavor := providerSpec.Get("value.flavor").String()
+		cpmsImage := providerSpec.Get("value.image").String()
+		cpmsSecurityGroups := make(map[string]struct{})
+		cpmsServerGroup := providerSpec.Get("value.serverGroupName").String()
+		cpmsSubnets := make(map[string]struct{})
+
+		for _, sg := range objects(providerSpec.Get("value.securityGroups")) {
+			securityGroupName := sg["name"].(string)
+			cpmsSecurityGroups[securityGroupName] = struct{}{}
+		}
+
+		for _, subnet := range objects(providerSpec.Get("value.networks[0].subnets")) {
+			subnetName := subnet.Get("filter.name").Str()
+			cpmsSubnets[subnetName] = struct{}{}
+		}
+
+		for _, machine := range masterMachines {
+			machineName := machine.Get("metadata.name").String()
+			g.By("Comparing the MachineSet spec with machine " + machineName)
+
+			machineFlavor := machine.Get("spec.providerSpec.value.flavor").String()
+			machineImage := machine.Get("spec.providerSpec.value.image").String()
+			machineSecurityGroups := make(map[string]struct{})
+			machineSubnets := make(map[string]struct{})
+			machineServerGroup := machine.Get("spec.providerSpec.value.serverGroupName").String()
+
+			for _, sg := range objects(machine.Get("spec.providerSpec.value.securityGroups")) {
+				securityGroupName := sg["name"].(string)
+				machineSecurityGroups[securityGroupName] = struct{}{}
+			}
+
+			for _, subnet := range objects(machine.Get("spec.providerSpec.value.networks[0].subnets")) {
+				subnetName := subnet.Get("filter.name").Str()
```

Comment:
Any particular reason to use `Str()` instead of `String()` here?

Reply:
Done